viewing_party/party.py

```diff
@@ -1,23 +1,212 @@
+    user_ratings_list = []
+    
+    for movie in user_data["watched"]:
+        user_ratings_list.append(movie["rating"])
+
+    try:
+        average_rating = sum(user_ratings_list) / len(user_data["watched"])
+        return average_rating
+    except(ZeroDivisionError):
+        return 0.0
+
+
+def get_most_watched_genre(user_data: dict) -> str:
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    user_genres_list = []
+    for movie in user_data["watched"]:
+        user_genres_list.append(movie["genre"])
+
+    user_most_watch_genre = max(set(user_genres_list), key=user_genres_list.count)    
+
+    return user_most_watch_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def user_wateched_movies(user_data: dict) -> set:
+    set_user_watched_movies = set()
+    for movie in user_data["watched"]: 
+        set_user_watched_movies.add(movie["title"])
+    
+    return set_user_watched_movies
+
+
+def friends_wateched_movies(user_data: dict) -> set:
+    set_friends_watched_movies = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            set_friends_watched_movies.add(movie["title"])
+    
+    return set_friends_watched_movies
+
+
+def remove_duplicate_values(list_of_values: list) -> list:
+    seen = set()
+    unique_values = []
+    for item in list_of_values:
+        t = tuple(item.items())
+        if t not in seen:
+            seen.add(t)
+            unique_values.append(item)
+            
+    return unique_values
+
+
+def get_unique_watched(user_data: dict) -> int:
+    user_watched_movies = user_wateched_movies(user_data)
+    friends_watched_movies = friends_wateched_movies(user_data)
+    user_unique_watched_movies = []
+    
+    set_user_unique_titles = user_watched_movies.difference(friends_watched_movies) # Set of movies(titles) only watched by user
+
+    for movie in user_data["watched"]:
+        if movie["title"] in set_user_unique_titles:
+            user_unique_watched_movies.append(movie)
+            
+    return user_unique_watched_movies
+
+
+def get_friends_unique_watched(user_data: dict) -> list: 
+    user_watched_movies = user_wateched_movies(user_data)
+    friends_watched_movies = friends_wateched_movies(user_data)
+    unique_recomended_movies = []
+     
+    set_friends_unique_movie_titles = friends_watched_movies.difference(user_watched_movies) #  Set of movies only watched by friends but not user
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in set_friends_unique_movie_titles:
+                if movie["title"] not in unique_recomended_movies:
+                    unique_recomended_movies.append(movie)
+    
+    unique_recomended_movies = remove_duplicate_values(unique_recomended_movies)
+
+    return unique_recomended_movies

```

Comment:
I love how you used the `set` here, however, it is super tricky when trying to find unique recommended movies. We can use `set` to compare immutable data types like the movie titles, however, it would be so nice just to compare the entire movie dictionaries! Because we only found the titles, this required you to have to iterate over the entire friend's list again in this function. If `friends_watched_movies` and `user_watched_movies` actually returned a list of dictionaries we would have been able to iterate through either list only once like so:

```python
for movie in friend_watched_movies:
    if movie not in user_watched_movies:
        unique_recommended_movies.append(movie)

# or

unique_recommended_movies = [movie in friends_watched_movies if movie not in user_watched_movies ]
```

Reply:
In other words, the `set` solution requires more iterations for finding the `unique_recommended_movies` (combined from the loops in this function and in the helper function).

Reply:
List comprehensions are way cleaner and shorter! I will be using them more often.